File: java/src/main/java/io/dure/array/FourNumberSum.java
package io.dure.array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FourNumberSum {

    public List<List<Integer>> fourNumberSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> quadruplets = new ArrayList<>();
        for (int i = 0; i < nums.length - 3; ) {
            for (int j = i + 1; j < nums.length - 2; ) {
                int left = j + 1;
                int right = nums.length - 1;
                while (left < right) {
                    int currentSum = nums[i] + nums[j] + nums[left] + nums[right];
                    if (currentSum == target) {
                        quadruplets.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        while (left < right && nums[left] == nums[++left]);
                        right--;
                    } else if (currentSum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
                while (j < nums.length - 2 && nums[j] == nums[++j]);
            }
            while (i < nums.length - 3 && nums[i] == nums[++i]);
        }
        return quadruplets;
    }
}

File: go/array/fourNumberSum.go
package array

import (
	"sort"
)

func fourNumberSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	var quadruplets [][]int
	for i := 0; i < len(nums)-3; {
		for j := i + 1; j < len(nums)-2; {
			left := j + 1
			right := len(nums) - 1
			for left < right {
				currentSum := nums[i] + nums[j] + nums[left] + nums[right]
				if currentSum == target {
					quadruplets = append(quadruplets, []int{nums[i], nums[j], nums[left], nums[right]})
					temp := nums[left]
					for left < right && temp == nums[left] {
						left++
					}
					right--
				} else if currentSum < target {
					left++
				} else {
					right--
				}
			}
			temp := nums[j]
			for j < len(nums)-2 && temp == nums[j] {
				j++
			}
		}
		temp := nums[i]
		for i < len(nums)-1 && temp == nums[i] {
			i++
		}

	}
	return quadruplets
}
